api/src/lib.rs

```diff
@@ -2589,4 +2594,1264 @@ pub extern "system" fn Java_com_horizen_sigproofnative_NaiveThresholdSigProof_na
         Ok(result) => if result { JNI_TRUE } else { JNI_FALSE },
         Err(_) => JNI_FALSE // CRYPTO_ERROR or IO_ERROR
     }
+}
+
+
+
+///////// COMMITMENT TREE
+#[no_mangle]
+pub extern "system" fn Java_com_horizen_commitmenttree_CommitmentTree_nativeInit(
+    _env: JNIEnv,
+    _class: JClass
+) -> jobject
+{
+
+    // Create new CommitmentTree Rust side
+    let commitment_tree = CommitmentTree::create();
+
+    // Create and return new CommitmentTree Java side
+    let commitment_tree_ptr: jlong = jlong::from(Box::into_raw(Box::new(commitment_tree)) as i64);
+
+    _env.new_object(_class, "(J)V", &[JValue::Long(commitment_tree_ptr)])
+        .expect("Should be able to create new CommitmentTree object")
+        .into_inner()
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_horizen_commitmenttree_CommitmentTree_nativeFreeCommitmentTree(
+    _env: JNIEnv,
+    _class: JClass,
+    _commitment_tree: *mut CommitmentTree
+)
+{
+    if _commitment_tree.is_null()  { return }
+    drop(unsafe { Box::from_raw(_commitment_tree) });
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_horizen_commitmenttree_CommitmentTree_nativeAddScCr(
+    _env: JNIEnv,
+    _commitment_tree: JObject,
+    _sc_id: jbyteArray,
+    _amount: jlong,
+    _pub_key: jbyteArray,
+    _withdrawal_epoch_length: jint,
+    _custom_data: jbyteArray,
+    _constant_nullable: jbyteArray,             // can be null if there is no constant
+    _cert_verification_key: jbyteArray,
+    _btr_verification_key_nullable: jbyteArray, // can be null if there is no key for BTRs
+    _csw_verification_key_nullable: jbyteArray, // can be null if there is no key for CSWs
+    _tx_hash: jbyteArray,
+    _out_idx: jint
+) -> jboolean
+{
+    let sc_id = {
+        let t = _env.convert_byte_array(_sc_id)
```

Comment:
this part of code (lines 2649 - 2657) actually is duplicated many times along the further code, so it could be implemented as an utility method like:
get_byte_array(_env: &JNIEnv, javaByteArray: &jbyteArray, length: usize) -> [u8]